src/workerd/io/trace-stream.c++

```diff
@@ -670,77 +698,134 @@ class TailStreamTargetBase: public rpc::TailStreamTarget::Server {
+  // the exported tailStream handler to be called, passing the onset event
+  // as the initial argument. If the tail stream wishes to continue receiving
+  // events for this invocation, it will return a handler in the form of an
+  // object or a function. If no handler is returned, the tail session is
+  // shutdown.
+  kj::Promise<void> handleOnset(Worker::Lock& lock,
+      IoContext& ioContext,
+      kj::ArrayPtr<tracing::TailEvent> events,
+      rpc::TailStreamTarget::TailStreamResults::Builder results) {
+    // There should be only a single onset event in this batch.
+    KJ_ASSERT(events.size() == 1 && events[0].event.is<tracing::Onset>(),
+        "Expected only a single onset event");
+    auto& event = events[0];
 
-// Returns the name of the handler function for this type of event.
-kj::Maybe<kj::StringPtr> getHandlerName(const tracing::TailEvent& event) {
-  KJ_SWITCH_ONEOF(event.event) {
-    KJ_CASE_ONEOF(_, tracing::Onset) {
-      return ONSET_STR;
-    }
-    KJ_CASE_ONEOF(_, tracing::Outcome) {
-      return OUTCOME_STR;
-    }
-    KJ_CASE_ONEOF(_, tracing::Hibernate) {
-      return HIBERNATE_STR;
-    }
-    KJ_CASE_ONEOF(_, tracing::SpanOpen) {
-      return SPANOPEN_STR;
-    }
-    KJ_CASE_ONEOF(_, tracing::SpanClose) {
-      return SPANCLOSE_STR;
+    // TODO(later): This assumes the tail worker is on the default export.
+    // Should we allow other named exports to provide the tailStream handler?
+    auto handler = KJ_REQUIRE_NONNULL(
+        lock.getExportedHandler("default"_kj, kj::mv(props), ioContext.getActor()),
+        "Failed to get handler to worker.");
+    StringCache stringCache;
+
+    jsg::Lock& js = lock;
+    auto target = jsg::JsObject(handler->self.getHandle(js));
+    v8::Local<v8::Value> maybeFn = target.get(js, "tailStream"_kj);
+
+    // If there's no actual tailStream handler, or if the tailStream export is
+    // something other than a function, we will emit a warning for the user
+    // then immediately return.
+    if (!maybeFn->IsFunction()) {
+      ioContext.logWarningOnce("A worker configured to act as a streaming tail worker does "
+                               "not export a tailStream() handler.");
+      results.setStop(true);
+      doneFulfiller->fulfill();
+      return kj::READY_NOW;
     }
-    KJ_CASE_ONEOF(mark, tracing::Mark) {
-      KJ_SWITCH_ONEOF(mark) {
-        KJ_CASE_ONEOF(_, tracing::DiagnosticChannelEvent) {
-          return DIAGNOSTICCHANNEL_STR;
-        }
-        KJ_CASE_ONEOF(_, tracing::Exception) {
-          return EXCEPTION_STR;
-        }
-        KJ_CASE_ONEOF(_, tracing::Log) {
-          return LOG_STR;
-        }
-        KJ_CASE_ONEOF(_, tracing::Return) {
-          return RETURN_STR;
+
+    // Invoke the tailStream handler function.
+    v8::Local<v8::Function> fn = maybeFn.As<v8::Function>();
+    v8::Local<v8::Value> obj = ToJs(js, event, stringCache);
+
+    try {
```

Comment:
This try-catch block seems to fix the issue being tested for in `tail-worker-test-invalid.wd-test` – it seems like we fail to setStop() when the tail worker throws an exception. I doubt it's the best way to address it, so input is welcome.